File: src/main/kotlin/com/malcolmcrum/adventofcode2019/Day15.kt
package com.malcolmcrum.adventofcode2019

import com.malcolmcrum.adventofcode2019.Droid.Entity.*
import com.malcolmcrum.adventofcode2019.Droid.Direction.*
import java.io.File
import kotlin.math.abs

class Droid {
    val knownMap = mutableMapOf<Tile, Entity>()
    var oxygen: Tile? = null
    var position = Tile(0, 0)
    lateinit var lastMove: Direction

    val move: () -> Long = {
        val (destination, path) = nearestUnknownTile() ?: throw MapExploredException()
        val direction = when (path.first()) {
            position + NORTH -> NORTH
            position + SOUTH -> SOUTH
            position + EAST -> EAST
            position + WEST -> WEST
            else -> error("Can't find first step from $position towards $destination in $path")
        }
        println("Found unknown tile at $destination. Moving $direction towards it, via ${path.first()}")
        lastMove = direction
        direction.code.toLong()
    }

    val update: (Long) -> Unit = {
        val output = Output.of(it.toInt())
        when (output) {
            Output.HIT_WALL -> {
                val wall = position + lastMove
                knownMap[wall] = WALL
                println("Hit wall at $wall")
            }
            Output.MOVED_STEP -> {
                position += lastMove
                knownMap[position] = EMPTY
                println("Moved to $position")
            }
            Output.MOVED_STEP_FOUND_OXYGEN -> {
                position += lastMove
                oxygen = position
                knownMap[position] = OXYGEN
                println("Found oxygen at $position")
            }
        }
    }

    private fun nearestUnknownTile(): Pair<Tile, List<Tile>>? {
        return findTile(position, position) { _, entity -> entity == null }
    }

    fun pathFrom(start: Tile, target: Tile): List<Tile> {
        return findTile(start, target) { tile, _ -> tile == oxygen}!!.second
    }

    // haha oh jeez. so wasteful
    fun maxDistanceFrom(start: Tile): Int {
        return knownMap.filter { it.value == EMPTY }
            .map { findTile(start, it.key) { tile, _ -> tile == it.key }!! }
            .map { it.second.size }
            .max()!!
    }

    // adapted from https://medium.com/@nicholas.w.swift/easy-a-star-pathfinding-7e6689c7f7b2
    fun findTile(start: Tile, target: Tile, test: (Tile, Entity?) -> Boolean): Pair<Tile, List<Tile>>? {
        val open = mutableMapOf(start to Node(start))
        val closed = mutableSetOf<Tile>()

        fun findMatchingTile(current: Node, direction: Direction): Pair<Tile, List<Tile>>? {
            val tile = current.tile + direction
            val entity = knownMap[tile]
            if (test.invoke(tile, entity)) {
                return Pair(tile, current.getPath().plus(tile).drop(1))
            } else if (entity == WALL || closed.contains(tile)) {
                // ignore it
            } else if (!open.contains(tile)) {
                open[tile] = Node(tile, current.cost + 1, current)
            } else {
                val existingCost = open[tile]!!.cost
                if (existingCost > current.cost + 1) {
                    open[tile] = Node(tile, current.cost + 1, current)
                }
            }
            return null
        }

        do {
            val current = open.minBy { (_, node) -> node.costTo(target) }!!.value
            open.remove(current.tile)
            closed.add(current.tile)
            val result = findMatchingTile(current, NORTH)
                ?: findMatchingTile(current, EAST)
                ?: findMatchingTile(current, SOUTH)
                ?: findMatchingTile(current, WEST)
            if (result != null) return result
        } while (open.isNotEmpty())
        return null
    }




    enum class Entity {
        WALL, EMPTY, OXYGEN
    }

    enum class Direction(val code: Int, val x: Int, val y: Int) {
        NORTH(1, 0, 1),
        SOUTH(2, 0, -1),
        WEST(3, -1, 0),
        EAST(4, 1, 0)
    }

    enum class Output(val code: Int) {
        HIT_WALL(0),
        MOVED_STEP(1),
        MOVED_STEP_FOUND_OXYGEN(2);

        companion object {
            fun of(code: Int) = values().find { it.code == code } ?: error("Output not recognized: $code")
        }
    }

    data class Node(val tile: Tile, val cost: Int = 0, var parent: Node? = null) {
        fun getPath(): List<Tile> {
            val parents = mutableListOf(tile)
            var parent = this.parent
            while (parent != null) {
                parents += parent.tile
                parent = parent.parent
            }
            parents.reverse()
            return parents
        }

        fun costTo(target: Tile): Int {
            return cost + abs(tile.first - target.first) + abs(tile.second - target.second)
        }
    }

}

class MapExploredException : Exception()

operator fun Tile.plus(direction: Droid.Direction): Tile {
    return Tile(first + direction.x, second + direction.y)
}

fun shortestPathToOxygen(): Int {
    val instructions = File("src/main/resources/input/day15.txt").readText()
        .split(",").map { it.toLong() }.toLongArray()
    val robot = Droid()
    try {
        Emulator5("Robot", instructions, robot.move, robot.update).run()
    } catch (e: MapExploredException) { // an awkward way to stop the computer...
        draw(robot.knownMap)
    }

    return robot.pathFrom(Tile(0, 0), robot.oxygen!!).size
}

fun draw(map: Map<Tile, Droid.Entity>) {
    var minX = Int.MAX_VALUE
    var maxX = Int.MIN_VALUE
    var maxY = Int.MIN_VALUE
    var minY = Int.MAX_VALUE
    for (tile in map.keys) {
        if (tile.first < minX) minX = tile.first
        if (tile.second < minY) minY = tile.second
        if (tile.first > maxX) maxX = tile.first
        if (tile.second > maxY) maxY = tile.second
    }
    val width = abs(maxX - minX) + 1
    val height = abs(maxY - minY) + 1
    val grid = array2dOfChar(width, height, ' ')
    for ((tile, entity) in map) {
        grid[tile.second - minY][tile.first - minX] = when(entity) {
            WALL -> '#'
            EMPTY -> '.'
            OXYGEN -> '@'
        }
    }
    grid.forEach { println(it) }
}

fun main() {
    val instructions = File("src/main/resources/input/day15.txt").readText()
        .split(",").map { it.toLong() }.toLongArray()
    val robot = Droid()
    try {
        Emulator5("Robot", instructions, robot.move, robot.update).run()
    } catch (e: MapExploredException) { // an awkward way to stop the computer...
        draw(robot.knownMap)
    }

    val steps = robot.pathFrom(Tile(0, 0), robot.oxygen!!).size
    println(steps)

    val minutesUntilFullOxygen = robot.maxDistanceFrom(robot.oxygen!!)

    println(minutesUntilFullOxygen)

}

File: python/day15.py
from __future__ import annotations

import sys
from enum import Enum
from typing import List, Set, Dict, Optional, Callable


class Entity(Enum):
    UNKNOWN = " "
    EMPTY = "."
    WALL = "#"
    OXYGEN = "@"


class Tile:
    def __init__(self, x: int, y: int):
        self.x = x
        self.y = y

    def __add__(self, direction: Direction) -> Tile:
        return Tile(self.x + direction.x, self.y + direction.y)

    def __repr__(self):
        return f"Tile{str(self)}"

    def __str__(self):
        return f"({self.x},{self.y})"

    def __hash__(self):
        return hash((self.x, self.y))

    def __eq__(self, other) -> bool:
        return self.x == other.x and self.y == other.y


class Droid:

    def __init__(self):
        self.position = Tile(0, 0)
        self.world = World()
        self.last_move: Optional[Direction] = None

    def move(self) -> Optional[int]:
        path_to_destination: List[Tile] = self.world.nearest_unexplored_tile(self.position)
        if not path_to_destination:
            return None
        direction: Direction = {self.position + d: d for d in Direction}[path_to_destination[1]]
        self.last_move = direction
        return direction.code

    def update(self, movement: int):
        movement_result = MovementResult(movement)
        if movement_result == MovementResult.HIT_WALL:
            wall = self.position + self.last_move
            self.world.add(Entity.WALL, wall)
        elif movement_result == MovementResult.MOVED_STEP:
            self.position = self.position + self.last_move
            self.world.add(Entity.EMPTY, self.position)
        elif movement_result == MovementResult.MOVED_STEP_FOUND_OXYGEN:
            self.position = self.position + self.last_move
            self.world.add(Entity.OXYGEN, self.position)


class World:

    def __init__(self):
        self.known_map: Dict[Tile, Entity] = {}

    def nearest_unexplored_tile(self, position: Tile) -> List[Tile]:
        return self.find_tile(position, position, lambda tile, entity: entity is Entity.UNKNOWN)

    def find_tile(self, start: Tile, destination: Tile, matches: Callable[[Tile, Entity], bool]) -> List[Tile]:
        open_tiles: Dict[Tile, Node] = {start: Node(start)}
        explored_tiles: Set[Tile] = set()

        def find_matching_tile(direction: Direction) -> Optional[List[Tile]]:
            tile: Tile = current.tile + direction
            entity: Entity = self.known_map.get(tile, Entity.UNKNOWN)
            cost: int = current.cost + 1
            if matches(tile, entity):
                return current.get_path() + [tile]
            elif entity is Entity.WALL or tile in explored_tiles:
                pass
            elif tile not in open_tiles:
                open_tiles[tile] = Node(tile, cost, current)
            else:
                existing_cost = open_tiles[tile].cost
                found_shorter_route = existing_cost > cost
                if found_shorter_route:
                    open_tiles[tile] = Node(tile, cost, current)
            return None

        while len(open_tiles) > 0:
            current = min(open_tiles.values(), key=lambda node: node.cost_to(destination))
            del open_tiles[current.tile]
            explored_tiles.add(current.tile)
            result: Optional[List[Tile], Direction] = find_matching_tile(Direction.NORTH) or \
                find_matching_tile(Direction.EAST) or \
                find_matching_tile(Direction.SOUTH) or \
                find_matching_tile(Direction.WEST)
            if result is not None:
                return result
        return []  # no tile found

    def add(self, entity: Entity, position: Tile):
        self.known_map[position] = entity


class Direction(Enum):
    NORTH = (1, 0, 1)
    EAST = (2, 1, 0)
    SOUTH = (3, 0, -1)
    WEST = (4, -1, 0)

    def __init__(self, code: int, x: int, y: int):
        self.code = code
        self.x = x
        self.y = y

    @staticmethod
    def of(code: int):
        return {direction.code: direction for direction in Direction}[code]


class MovementResult(Enum):
    HIT_WALL = 0
    MOVED_STEP = 1
    MOVED_STEP_FOUND_OXYGEN = 2


class Node:
    def __init__(self, tile: Tile, cost: int = 0, parent: Optional[Node] = None):
        self.tile = tile
        self.parent = parent
        self.cost = cost

    def get_path(self):
        parents: List[Tile] = [self.tile]
        parent = self.parent
        while parent is not None:
            parents.append(parent.tile)
            parent = parent.parent
        parents.reverse()
        return parents

    def cost_to(self, target: Tile) -> int:
        return self.cost + abs(self.tile.x - target.x) + abs(self.tile.y - target.y)


def draw(world: Dict[Tile, Entity]):
    min_x, min_y = sys.maxsize, sys.maxsize
    max_x, max_y = -sys.maxsize, -sys.maxsize
    for tile in world.keys():
        if tile.x < min_x:
            min_x = tile.x
        if tile.y < min_y:
            min_y = tile.y
        if tile.x > max_x:
            max_x = tile.x
        if tile.y > max_y:
            max_y = tile.y
    for y in range(min_y, max_y + 1):
        for x in range(min_x, max_x + 1):
            entity = world.get(Tile(x, y), " ")
            print(Entity(entity).value, end="")
        print()


file = open("map.txt", "r")
if file.mode == "r":
    contents = file.readlines()
    droid = Droid()
    droid.position = Tile(2, 2)
    while True:
        command: Optional[int] = droid.move()
        if command is None:
            draw(droid.world.known_map)
            exit(0)
        destination: Tile = droid.position + Direction.of(command)
        result: str = contents[destination.y][destination.x]
        if result == Entity.EMPTY.value:
            droid.update(MovementResult.MOVED_STEP.value)
        elif result == Entity.WALL.value:
            droid.update(MovementResult.HIT_WALL.value)
        elif result == Entity.OXYGEN.value:
            droid.update(MovementResult.MOVED_STEP_FOUND_OXYGEN.value)
